enhancements/ingress/tunable-router-buffer-sizes.md

```diff
@@ -0,0 +1,165 @@
+
+The IngressController API is extended by adding an optional `HeaderBuffer` field with
+type `*IngressControllerHeaderBuffer` to `IngressControllerSpec`:
+
+```go
+type IngressControllerSpec struct {
+	// ...
+
+	// HeaderBuffer defines parameters for buffer size values. If this
+	// field is empty, the default values set within the IngressController
+	// will be used.
+	//
+	// +optional
+	HeaderBuffer *IngressControllerHeaderBuffer `json:"headerBuffer,omitempty"`
+}
+```
+
+The `IngressControllerHeaderBuffer` type has 2 `int` fields, one for
+`HeaderBufferSize` and one for `HeaderBufferMaxRewrite`.
+
+```go
+type IngressControllerHeaderBuffer struct {
+	// buffersize describes how much memory should be reserved
+	// (in bytes) for IngressController connection sessions.
+	//
+	// +kubebuilder:validation:required 
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferSize int `json:"headerBufferSize,omitempty"`
+
+	// buffermaxrewrite describes how much memory should be reserved
+	// (in bytes) from HeaderBufferSize for HTTP header rewriting
+	// and appending.
+	//
+	// +kubebuilder:validation:required
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferMaxRewrite int `json:"headerBufferMaxRewrite,omitempty"`
+}
+```
+
+Proper environment variables will be added to the Router deployment to allow for the `HeaderBufferSize` and `HeaderBufferMaxRewrite` values to be
+populated in the HAProxy configuration template.
+
+The [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) will need to be modified to handle both the API changes
+and setting the new Router environment variables.
+
+Integration tests will also need to be written to ensure that the new IngressController API fields work as intended (and that the router
+respects buffer limits without degrading).
+
+### User Stories
+
+#### As a Cluster Administrator, I want to edit the HAProxy configuration by setting "tune.maxrewrite" and "tune.bufsize" so that the Router can accommodate requests with large header sizes
+
+The necessary Router and API changes will be made to allow cluster administrators to tweak these settings on a per-IngressController level.
+
+### Implementation Details/Notes/Constraints
+
+Header buffer size values should be validated to make sure they are at least sane. For example, `HeaderBufferSize` values of `0` or `MAXINT` should most likely
+be rejected.
+
+### Risks and Mitigations
+
+Cluster administrators will have the ability to set IngressController buffer sizes that break cluster ingress.
+It is the responsibility of the administrator to modify the HAProxy buffer values responsibly and appropriately.
+Ideally, the buffer size values would only be modified if absolutely needed.
+
+## Design Details
+
+### Open Questions
+
+How often will cluster administrators take advantage of this enhancement?
+Or, in other words, how sufficient are the current HAProxy buffer default values?
```

Comment:
We don't have a way to determine now what administrators will do, but once this feature is shipped and potentially in use, will we be able to use metrics to see if these values have been modified?

Reply:
Hm, im not sure how we would add a metric to track this sort of data, since we would only want to record whether or not a cluster administrator has changed the default values, not how often they have, if they chose to do so. I am also not sure if this sort of metric would be telemetry-worthy.

Reply:
I suppose we could have a gauge metric that just tracks the number of ingress controllers with this API field set, but again, I wonder how useful that data actually is (and if we would really be interested in collecting it form customers).

Reply:
In the past, we've discussed adding a similar metric for DNS (see https://github.com/openshift/cluster-dns-operator/pull/142), which would be useful for quickly checking if a cluster is using the feature, or how many clusters in a fleet are using the feature.